docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,557 @@
+# ADR 014: Bootstrap from previous peers
+
+## Changelog
+
+* 2023-05-04: initial draft
+* 2023-05-04: update peer selection with new simpler solution + fix language
+* 2023-10-04: propose a second approach
+* 2023-11-04: propose a third approach
+
+## Context
+
+Celestia node relies on a set of peer to peer protocols that assume a decentralized network, where nodes are connected to each other and can exchange data.
+
+Currently, when a node joins/comes up on the network, it relies, by default, on hardcoded bootstrappers for 3 things:
+
+1. initializing its store with the first (genesis) block
+2. requesting the network head to which it sets its initial sync target, and
+3. bootstrapping itself with some other peers on the network which it can utilise for its operations.
+
+This is a centralization bottleneck as it happens both during initial start-up and for any re-starts.
+
+In this ADR, we wish to aleviate this problem by allowing nodes to bootstrap from previously seen peers. This is a simple solution that allows the network alleviate the start-up centralisation bottleneck such that nodes can join the network without the need for hardcoded bootstrappers up re-start
+
+(_Original issue available in references_)
+
+## Decision
+
+Periodically store the addresses of previously seen peers in a key-value database, and allow the node to bootstrap from these peers on start-up.
+
+## Design
+
+## Approach A: Hook into `peerTracker` from `libhead.Exchange` and react to its internal events
+
+</br>
+
+In this approach, we will rely on `peerTracker`'s scoring mechanism that tracks the fastest connected nodes, and blocks malicious ones. To make use of this functionality, we track two important internal events from `peerTracker` in `libhead.Exchange` for peers selection and persistence purposes, and that is by proposing a design that allows access to `peerTracker`'s list of good peers and bad peers, and opening the design space to integrate peer selection and storage logic.
+
+## Implementation
+
+</br>
+
+### What systems will be affected?
+
+* `peerTracker` and `libhead.Exchange` from `go-header`
+* `newInitStore` in `nodebuilder/header.go`
+
+### Storage
+
+We will use a badgerDB datastore to store the addresses of previously seen good peers. The database will be stored in the `data` directory, and will have a store prefix `good_peers`. The database will have a single key-value pair, where the key is `peers` and the value is a `[]PeerInfo`
```

Comment:
This will be the only key in the datastore that receives updates, filling the value log. Because it's not updated very often and the value is very large, it won't cause bloat, but we should still note this. It probably means we should keep GC enabled.